src/toolkit/res_map.h

```diff
@@ -0,0 +1,191 @@
+#ifndef CYCLUS_SRC_TOOLKIT_RES_MAP_H_
+#define CYCLUS_SRC_TOOLKIT_RES_MAP_H_
+
+#include <iomanip>
+#include <limits>
+#include <map>
+#include <set>
+
+#include "cyc_arithmetic.h"
+#include "cyc_limits.h"
+#include "error.h"
+#include "product.h"
+#include "material.h"
+#include "resource.h"
+#include "res_manip.h"
+
+namespace cyclus {
+namespace toolkit {
+
+/// ResMap container for the management of resources. It allows you to associate
+/// keys with individual resources. The keys are often strings or ints and the 
+/// ResMap enables you to add whatever semantic meaning that you want to these keys.
+/// Like the ResBuf class, ResMap is able to keep track of its total quantity of 
+/// resources and has a maximum capacity limit. Overfilled ResMaps will throw a 
+/// ValueError when their quantity() is next computed.
+///
+/// Typically, a ResMap will be a member variable of an archetype class.
+/// Resources can be added, removed, and retrieved from it as needed.
+///
+/// @code
+/// class HasResMapAgent : public cyclus::Facility {
+///  public:
+///   Tick() {
+///     cyclus::Material::Ptr best = inventory_["best"];
+///     double invsize = inventory_.quantity();
+///   }
+///
+///  private:
+///   cyclus::toolkit::ResMap<std::string, cyclus::Material> inventory_;
+/// };
+/// @endcode
+template <class K, class R>
+class ResMap {
+ public:
+  ResMap() : capacity_(INFINITY), dirty_quantity_(true), quantity_(0) {
+    Warn<EXPERIMENTAL_WARNING>("ResMap is experimental and its API may be "
+                               "subject to change");
+  }
+
+  virtual ~ResMap() {}
+
+  typedef typename std::map<K, typename R::Ptr> map_type;
```

Comment:
You need to use Resource::obj_id's here.  Pointers are not the consistent from snapshot/init.  So you will also need to add another map from obj_id to resource_ptr.  The InitInv codegen for ResMap will need to populate the obj_id->ptr map while the InitFrom(db) codegen will populate the K->obj_id map.